filecoin-proofs/src/api/update.rs

```diff
@@ -203,6 +209,84 @@ pub fn encode_into<Tree: 'static + MerkleTreeTrait<Hasher = TreeRHasher>>(
     })
 }
 
+/// Decodes a range of data with the given sector key.
+///
+/// This function is similar to [`decode_from`], the difference is that it operates directly on the
+/// given file descriptions. The current position of the file descriptors is where the decoding
+/// starts, i.e. you need to seek to the intended offset before you call this function. The
+/// `nodes_offset` is the node offset relative to the beginning of the file. This information is
+/// needed in order to do the decoding correctly. The `nodes_count` is the total number of nodes
+/// within the file. The `num_nodes` defines how many nodes will be decoded, starting from the
+/// current position.
+#[allow(clippy::too_many_arguments)]
+pub fn decode_from_range<R: Read, S: Read, W: Write>(
+    nodes_count: usize,
+    comm_d: Commitment,
+    comm_r: Commitment,
+    input_data: R,
+    sector_key_data: S,
+    output_data: &mut W,
+    nodes_offset: usize,
+    num_nodes: usize,
+) -> Result<()> {
+    let comm_d_domain = TreeDDomain::try_from_bytes(&comm_d[..])?;
+    let comm_r_domain = TreeRDomain::try_from_bytes(&comm_r[..])?;
+    let phi = phi(&comm_d_domain, &comm_r_domain);
+    let rho_invs = Rhos::new_inv_range(&phi, nodes_count, nodes_offset, num_nodes);
+
+    let bytes_length = num_nodes * NODE_SIZE;
+
+    let input_iter = ChunkIterator::new(input_data);
+    let sector_key_iter = ChunkIterator::new(sector_key_data);
+    let chunk_size = input_iter.chunk_size();
+
+    for (chunk_index, (input_chunk_result, sector_key_chunk_result)) in
+        input_iter.zip(sector_key_iter).enumerate()
+    {
+        let chunk_offset = chunk_index * chunk_size;
+
+        // The end of the intended decoding range was reached.
+        if chunk_offset > bytes_length {
+            break;
+        }
+
+        let input_chunk = input_chunk_result?;
+        let sector_key_chunk = sector_key_chunk_result?;
```

Comment:
I think adding `context` to these read results (at least "input_data/sector_key read failure") would be a useful for determining where exactly an io error occurred.